### PR TITLE
Assert is not a function

### PR DESCRIPTION
**Pull request by EmilyBourne**
Assert is not a function. Remove all brackets that handle an assertion like a function call. Docs are also added and dead code is removed from the Errors classes. In particular the function `Errors.set_target` is modified to remove a parameter which only ever takes one argument

### CODE REVIEW

File: pyccel/ast/builtins.py [hunk 1]
```diff
@@ -1145,7 +1150,7 @@ def __call__(self, *args):
         """ Returns the expression with the arguments replaced with
         the calling arguments
         """
-        assert(len(args) == len(self.variables))
+        assert len(args) == len(self.variables)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/builtins.py:1153]: /bot accept
Issue opened : #1849

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 2]
```diff
@@ -3554,35 +3554,35 @@ def create_zeros_like_val_complex128():
...
     assert matching_types(f_int_complex(), create_zeros_like_val_complex())
 
     f_real_int32   = epyccel(create_zeros_like_val_int32, language = language)
-    assert(     f_real_int32()       ==      create_zeros_like_val_int32())
+    assert      f_real_int32()       ==      create_zeros_like_val_int32()
     assert matching_types(f_real_int32(), create_zeros_like_val_int32())
 
     f_real_float32   = epyccel(create_zeros_like_val_float32, language = language)
-    assert(isclose(     f_real_float32()    ,      create_zeros_like_val_float32(), rtol=RTOL, atol=ATOL))
+    assert isclose(     f_real_float32()    ,      create_zeros_like_val_float32(), rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3573]:
```suggestion
    assert isclose(f_real_float32(), create_zeros_like_val_float32(), rtol=RTOL, atol=ATOL)
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 3]
```diff
@@ -3554,35 +3554,35 @@ def create_zeros_like_val_complex128():
...
     assert matching_types(f_real_float32(), create_zeros_like_val_float32())
 
     f_real_float64   = epyccel(create_zeros_like_val_float64, language = language)
-    assert(isclose(     f_real_float64()    ,      create_zeros_like_val_float64(), rtol=RTOL, atol=ATOL))
+    assert isclose(     f_real_float64()    ,      create_zeros_like_val_float64(), rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3577]:
```suggestion
    assert isclose(f_real_float64(), create_zeros_like_val_float64(), rtol=RTOL, atol=ATOL)
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 4]
```diff
@@ -3554,35 +3554,35 @@ def create_zeros_like_val_complex128():
...
     assert matching_types(f_real_float64(), create_zeros_like_val_float64())
 
     f_real_complex64   = epyccel(create_zeros_like_val_complex64, language = language)
-    assert(isclose(     f_real_complex64()  ,      create_zeros_like_val_complex64(), rtol=RTOL, atol=ATOL))
+    assert isclose(     f_real_complex64()  ,      create_zeros_like_val_complex64(), rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3581]:
```suggestion
    assert isclose(f_real_complex64(), create_zeros_like_val_complex64(), rtol=RTOL, atol=ATOL)
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 5]
```diff
@@ -3554,35 +3554,35 @@ def create_zeros_like_val_complex128():
...
     assert matching_types(f_real_complex64(), create_zeros_like_val_complex64())
 
     f_real_complex128   = epyccel(create_zeros_like_val_complex128, language = language)
-    assert(isclose(     f_real_complex128() ,      create_zeros_like_val_complex128(), rtol=RTOL, atol=ATOL))
+    assert isclose(     f_real_complex128() ,      create_zeros_like_val_complex128(), rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3585]:
```suggestion
    assert isclose(f_real_complex128(), create_zeros_like_val_complex128(), rtol=RTOL, atol=ATOL)
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 6]
```diff
@@ -3703,20 +3703,20 @@ def create_zeros_like_3_val():
 
     f1_shape = epyccel(create_zeros_like_1_shape, language = language)
     f1_val   = epyccel(create_zeros_like_1_val, language = language)
-    assert(     f1_shape() ==      create_zeros_like_1_shape()      )
-    assert(     f1_val()   ==      create_zeros_like_1_val()        )
+    assert      f1_shape() ==      create_zeros_like_1_shape()
+    assert      f1_val()   ==      create_zeros_like_1_val()
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3707]:
```suggestion
    assert f1_shape() == create_zeros_like_1_shape()
    assert f1_val()   == create_zeros_like_1_val()
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 7]
```diff
@@ -3703,20 +3703,20 @@ def create_zeros_like_3_val():
...
     assert matching_types(f1_val(), create_zeros_like_1_val())
 
     f2_shape = epyccel(create_zeros_like_2_shape, language = language)
     f2_val   = epyccel(create_zeros_like_2_val, language = language)
-    assert(     f2_shape() ==      create_zeros_like_2_shape()      )
-    assert(isclose(     f2_val()  ,      create_zeros_like_2_val()        , rtol=RTOL, atol=ATOL))
+    assert      f2_shape() ==      create_zeros_like_2_shape()
+    assert isclose(     f2_val()  ,      create_zeros_like_2_val()        , rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3713]:
```suggestion
    assert f2_shape() == create_zeros_like_2_shape()
    assert isclose(f2_val(), create_zeros_like_2_val(), rtol=RTOL, atol=ATOL)
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 8]
```diff
@@ -3703,20 +3703,20 @@ def create_zeros_like_3_val():
...
     assert matching_types(f2_val(), create_zeros_like_2_val())
 
     f3_shape = epyccel(create_zeros_like_3_shape, language = language)
     f3_val   = epyccel(create_zeros_like_3_val, language = language)
-    assert(     f3_shape() ==      create_zeros_like_3_shape()      )
-    assert(isclose(     f3_val()  ,      create_zeros_like_3_val()        , rtol=RTOL, atol=ATOL))
+    assert      f3_shape() ==      create_zeros_like_3_shape()
+    assert isclose(     f3_val()  ,      create_zeros_like_3_val()        , rtol=RTOL, atol=ATOL)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3719]:
```suggestion
    assert f3_shape() == create_zeros_like_3_shape()
    assert isclose(f3_val(), create_zeros_like_3_val(), rtol=RTOL, atol=ATOL)
```

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good to me.

I am suggesting many small changes which you can apply all together by clicking on "Add suggestion to batch".

**Comment by EmilyBourne** [comment]
> Looks good to me.
> 
> I am suggesting many small changes which you can apply all together by clicking on "Add suggestion to batch".

I nearly went through and removed all this whitespace but when I checked who'd added it, it was you  😅

**Comment by yguclu** [comment]
> > Looks good to me.
> > 
> > I am suggesting many small changes which you can apply all together by clicking on "Add suggestion to batch".
> 
> I nearly went through and removed all this whitespace but when I checked who'd added it, it was you  😅

I might have used some whitespace to align instructions a long time ago, but many commits happened since then and now the code is a mess.

**Comment by yguclu** [comment]
Further, not every suggestion is about whitespace only...

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 9]
```diff
@@ -1531,35 +1531,35 @@ def create_full_val_real_complex128(val : 'float'):
     val_float = rand()*100
 
     f_int_int   = epyccel(create_full_val_int_int, language = language)
-    assert(     f_int_int(val_int)        ==      create_full_val_int_int(val_int))
+    assert      f_int_int(val_int)        ==      create_full_val_int_int(val_int)
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:1534]:
```suggestion
    assert f_int_int(val_int) == create_full_val_int_int(val_int)
```
Not here?

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 10]
```diff
@@ -2243,27 +2243,27 @@ def create_array_vals_2d():
     m = randint(10)
     p = randint(5)
     f_1d = epyccel(create_array_size_1d, language = language)
-    assert( f_1d(n)       == create_array_size_1d(n)      )
+    assert  f_1d(n)       == create_array_size_1d(n)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Not:
```suggestion
    assert f_1d(n) == create_array_size_1d(n)
```
?

**Comment by EmilyBourne** [comment]
@yguclu I have applied all your suggestions. I didn't see any non-whitespace changes though.

Unfortunately it seems that GitHub struggles when you apply this many suggestions in one go so I have gone through the changes individually to close the conversations that have been applied. I didn't leave a comment with the commit though.

I think there are probably many other similar whitespace messes in this file (this is the other reason why I didn't do this earlier as I didn't want to make the diff bigger). Do you want me to fix this anyway? I can use vim shortcuts to do this:
```
%s/,/, /g
%s/,  \+/, /g
%s/( \+/(/g
%s/ \+)/)/g
%s/  \+=/ =/g
%s/=  \+/= /g
```

But this will unalign the few places where the alignment still works. E.g:
```
    f3_shape = epyccel(create_ones_like_3_shape, language = language)
    f3_val   = epyccel(create_ones_like_3_val, language = language)
```

On the other hand this should prevent messes like this reappearing in the future

**Comment by EmilyBourne** [comment]
> > @yguclu I have applied all your suggestions. I didn't see any non-whitespace changes though.
> 
> In a few places tuples of integers were compared using
> 
> ```python
> assert all(isclose(a, b))
> ```
> 
> instead of simply writing
> 
> ```python
> assert a == b
> ```
> 
> I suggested a correction for those I could spot.

A bare `is_close` is not a good test even for floats. Tolerances should be specified. I will go through and weed these out

**Comment by yguclu** [comment]
> I think there are probably many other similar whitespace messes in this file (this is the other reason why I didn't do this earlier as I didn't want to make the diff bigger). Do you want me to fix this anyway? I can use vim shortcuts to do this:
> 
> ```
> %s/,/, /g
> %s/,  \+/, /g
> %s/( \+/(/g
> %s/ \+)/)/g
> %s/  \+=/ =/g
> %s/=  \+/= /g
> ```
> 
> But this will unalign the few places where the alignment still works. E.g:
> 
> ```
>     f3_shape = epyccel(create_ones_like_3_shape, language = language)
>     f3_val   = epyccel(create_ones_like_3_val, language = language)
> ```
> 
> On the other hand this should prevent messes like this reappearing in the future

I think we have done enough for this PR 😅

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 9 above)
Comment (yguclu) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:1534]: Yes, here too!

**Comment by EmilyBourne** [comment]
> A bare `is_close` is not a good test even for floats. Tolerances should be specified. I will go through and weed these out

You found all but one of them.
Done : 7b01e13eac9f95162006133f07726f03b1e291f9

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 9 above)
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Done ad907a8422685d76810932d72f4959706fda2342

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Extra space removed: 4c5fe1759

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 11]
```diff
@@ -2243,27 +2243,27 @@ def create_array_vals_2d():
...
 
     f_2d = epyccel(create_array_size_2d, language = language)
-    assert( f_2d(n, m)    == create_array_size_2d(n, m)   )
+    assert  f_2d(n, m)    == create_array_size_2d(n, m)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Extra space removed: 4c5fe1759

**Review by yguclu — APPROVED** [review]

**Pull request merged**